File: solutions/elixir/031-shortest-path-to-get-food.ex
# Problem 31: Shortest Path to Get Food (LeetCode 1730)
# Difficulty: Med
# Language: Elixir
# 
defmodule ShortestFood do
  def get_food(grid) do
    rows = length(grid); cols = length(hd(grid))
    arr = grid |> Enum.with_index() |> Enum.flat_map(fn {row, i} ->
      row |> Enum.with_index() |> Enum.map(fn {c, j} -> {{i, j}, c} end)
    end) |> Map.new()

    {sr, sc} = Enum.find_value(arr, fn {{r, c}, v} -> if v == "*", do: {r, c} end)
    dirs = [{0,1},{0,-1},{1,0},{-1,0}]
    bfs([{sr, sc, 0}], MapSet.new([{sr, sc}]), arr, rows, cols, dirs)
  end

  defp bfs([], _vis, _arr, _rows, _cols, _dirs), do: -1
  defp bfs(queue, vis, arr, rows, cols, dirs) do
    {next, vis, found} = Enum.reduce(queue, {[], vis, nil}, fn
      _, {nx, vs, {:found, d}} -> {nx, vs, {:found, d}}
      {r, c, d}, {nx, vs, nil} ->
        Enum.reduce(dirs, {nx, vs, nil}, fn
          _, {nx, vs, {:found, d}} -> {nx, vs, {:found, d}}
          {dr, dc}, {nx, vs, nil} ->
            {nr, nc} = {r + dr, c + dc}
            if nr >= 0 and nr < rows and nc >= 0 and nc < cols and
               Map.get(arr, {nr, nc}) != "X" and not MapSet.member?(vs, {nr, nc}) do
              if Map.get(arr, {nr, nc}) == "#", do: {nx, vs, {:found, d + 1}},
              else: {[{nr, nc, d + 1} | nx], MapSet.put(vs, {nr, nc}), nil}
            else
              {nx, vs, nil}
            end
        end)
    end)
    case found do
      {:found, d} -> d
      nil -> bfs(Enum.reverse(next), vis, arr, rows, cols, dirs)
    end
  end
end

File: solutions/typescript/031-shortest-path-to-get-food.ts
/**
 * Problem 31: Shortest Path to Get Food (LeetCode 1730)
 * Difficulty: Med
 * Language: TypeScript
 */
function getFood(grid: string[][]): number {
  const rows = grid.length, cols = grid[0].length;
  const start = grid.reduce<[number, number]>((acc, row, i) => {
    const j = row.indexOf('*');
    return j !== -1 ? [i, j] : acc;
  }, [0, 0]);

  const dirs = [[0,1],[0,-1],[1,0],[-1,0]];

  const bfs = (queue: [number, number, number][], visited: Set<string>): number => {
    if (queue.length === 0) return -1;
    const next: [number, number, number][] = [];
    for (const [r, c, d] of queue) {
      for (const [dr, dc] of dirs) {
        const nr = r + dr, nc = c + dc;
        const key = `${nr},${nc}`;
        if (nr >= 0 && nr < rows && nc >= 0 && nc < cols &&
            grid[nr][nc] !== 'X' && !visited.has(key)) {
          if (grid[nr][nc] === '#') return d + 1;
          visited.add(key);
          next.push([nr, nc, d + 1]);
        }
      }
    }
    return bfs(next, visited);
  };

  const visited = new Set<string>([`${start[0]},${start[1]}`]);
  return bfs([[start[0], start[1], 0]], visited);
}
